### ISSUE TITLE
Bug in calculation of p-values?

### ISSUE THREAD

**Issue by tajuma** [arcetri/sts#8]
I was cross/double checking some of the statistics of my randomness tests but could not get the p-values reported in finalAnaysisReport to agree with my calculation. The difference seems to be that sts rounds the sample size divided by 10 which then obviously gives different results than without rounding when sample size is not divisible by 10.

For example, one result from

```
% ./tools/generators 9 64 > sha1.bin
% ./sts -O -F r -i 64 sha1.bin
 
% head -8 finalAnalysisReport.txt | tail -3
 C1  C2  C3  C4  C5  C6  C7  C8  C9 C10  P-VALUE  PROPORTION  STATISTICAL TEST
------------------------------------------------------------------------------
  6   8   6   6  16   4   3   5   5   5  0.017912     61/64      Frequency

% python3
>>> from scipy.special import gammaincc
>>> import numpy as np

>>> obs = np.array([6,   8,   6,   6,  16,   4,   3,   5,   5,   5])

>>> chi_float = sum(((obs - 6.4)**2) / 6.4)
>>> gammaincc(9/2, chi_float/2)
0.029796344939787778

>>> chi_int = sum(((obs - 6)**2) / 6)
>>> gammaincc(9/2, chi_int/2)
0.01791240452984323
```
shows that sts reports chi_int based p-value due to integer division at [1]. 
Statistics is not my strong point ;-) but I assume it should be something like 

`expCount = (double)sampleCount / state->tp.uniformity_bins;`

to get the correct p-value of 0.029796.

[1] [https://github.com/arcetri/sts/blob/master/src/tests/frequency.c#L673](https://github.com/arcetri/sts/blob/master/src/tests/frequency.c#L673)

**Comment by lcn2** [comment]
Hello @tajuma,

You make a very good point that performing an integer division calculation `sampleCount / state->tp.uniformity_bins` when computing the dobule `expCount` is not correct: especially when `sampleCount` is small.

While `sampleCount` should not be small for the purpose of generating meaningful results, some people simply don't collect enough data and run the tool anyway.  For past results where the `sampleCount` was large, the integer rounding impact was minimized.  That's **not** an excuse intended to justify that integer division: just an observation.  

We are inclined to agree with your assessment.

Please consider, if you will, making a pull request making the correction of casting the `sampleCount` to double before dividing.

**Comment by lcn2** [comment]
Fixed in commit 800e8d7dec547e480319ba48f5d18c726432ad07

**THANK YOU** @tajuma .. we have also gratefully acknowledged your contribution in the commit log.
